File: leetcode/powx-n.cpp
/* -------------------------------------------------------------------------- */
/*                    https://leetcode.com/problems/powx-n                    */
/* -------------------------------------------------------------------------- */

#include <iostream>

using namespace std;

class Solution {
   public:
    double myPow(double x, int n) {
        double ans = 1;
        while (n != 0) {
            // If the power is odd you have to
            // multiply the candidate once to the answer
            if (n & 1 == 1) {
                if (n > 0) {
                    ans = ans * x;
                } else {
                    ans = ans / x;
                }
            }

            x = x * x;  // Then square the cadidate
            n = n / 2;  // Power becomes half
        }
        return ans;
    }
};

File: leetcode/powx-n.rs
/* -------------------------------------------------------------------------- */
/*                    https://leetcode.com/problems/powx-n                    */
/* -------------------------------------------------------------------------- */

fn my_pow(x: f64, n: i32) -> f64 {
    let mut ans: f64 = 1.0;
    let mut num = x;
    let mut p = n;
    while p != 0 {
        if p & 1 == 1 {
            if p > 0 {
                ans = ans * num;
            } else {
                ans = ans / num;
            }
        }

        num = num * num;
        p = p / 2;
    }
    ans
}

fn main() {
    println!("{}", my_pow(2.0, 10));
}
